presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PropertyDerivations.java

```diff
@@ -395,15 +395,17 @@ public ActualProperties visitJoin(JoinNode node, List<ActualProperties> inputPro
                     buildProperties = buildProperties.translate(column -> filterIfMissing(node.getOutputSymbols(), column));
 
                     return ActualProperties.builderFrom(buildProperties.translate(column -> filterIfMissing(node.getOutputSymbols(), column)))
+                            // TODO: some probe properties could be retained (e.g: grouping)
```

Comment:
@sopel39 @findepi I assume that right join is the same as left join but with probe and build sides reversed. I expect both joins to preserve properties in the same way after swapping build and probe.

Furthermore, I believe the following to be true:

* LEFT join preserves ALL properties of the PROBE side and no properties of the build side
* RIGHT join preserves ALL properties of the BUILD side and no properties of the probe side
* FULL join doesn't preserve any properties of any side

Left join inserts nulls at random places for build columns, hence breaking build properties. Right join inserts nulls at random places for probe columns, hence breaking probe properties. Full join inserts null at random places for build and probe columns, hence breaking probe and build properties.

Do you agree?

Reply:
> * LEFT join preserves ALL properties of the PROBE side and no properties of the build side

correct

> * RIGHT join preserves ALL properties of the BUILD side and no properties of the probe side
> * FULL join doesn't preserve any properties of any side

No, right and full join execute by streaming over the table on the probe side and emitting positive matches as they go. However, at the end they will output a row containing null for columns on the left side for every row on the right side that didn't match.

Reply:
> I assume that right join is the same as left join but with probe and build sides reversed. I expect both joins to preserve properties in the same way after swapping build and probe.

@mbasmanova from relations (sets of rows) perspective -- yes.

implementation-wise -- no. In LEFT case, the 'OUTER' rows are "inline", emitted as we go. In RIGHT case, the 'OUTER' rows are emitted at the end, so reshuffling occurs.
And derived properties physical, so they describe the underlying implementation and need to reflect that.

Reply:
@sopel39 @martint @findepi Folks, thanks for clarifying. Is the following true?

-- SPILL is OFF

* LEFT join preserves ALL properties of the PROBE side and no properties of the build side
* RIGHT and FULL joins preserve grouping and sorted-nulls-last properties of the probe side

-- SPILL is ON

* LEFT join preserves constant properties of the probe side
* RIGHT and FULL join don't preserve any properties

Reply:
@mbasmanova

* SPILL is OFF
  * LEFT/INNER -- i agree
  * RIGHT/FULL -- i agree
* SPILL is ON
  * LEFT/INNER -- i agree. (Arbitrary reshuffling can occur, constants are preserved)
  * RIGHT/FULL -- spill is not implemented for RIGHT/FULL (and `spillPossible` would return `false` for them), so we cannot reason about physical properties of produced output.

Reply:
@findepi Piotr, thanks. @martint Martin, could you take another look? If we all agree, I'll make a PR to update `PropertyDerivations` accordingly.

Reply:
@mbasmanova, that sounds about right.

Reply:
>-- SPILL is ON
>
>    LEFT join preserves constant properties of the probe side

Spilling join not preserving grouping could be a problem because then streaming aggregation will never happens for correlated subqueries and spill enabled.

I wonder how this could be addressed, e.g: make sure spill doesn't happen mid-group.